tests/nativeapp/test_artifacts.py

```diff
@@ -44,6 +46,512 @@ def dir_structure(path: Path, prefix="") -> List[str]:
     return parts
 
 
+@pytest.fixture
+def bundle_map():
+    project_files = {
+        "snowflake.yml": "# empty",
+        "README.md": "# Test Project",
+        "app/setup.sql": "-- empty",
+        "app/manifest.yml": "# empty",
+        "src/snowpark/main.py": "# empty",
+        "src/snowpark/a/file1.py": "# empty",
+        "src/snowpark/a/file2.py": "# empty",
+        "src/snowpark/a/b/file3.py": "# empty",
+        "src/snowpark/a/b/file4.py": "# empty",
+        "src/snowpark/a/c/file5.py": "# empty",
+        "src/streamlit/main_ui.py": "# empty",
+        "src/streamlit/helpers/file1.py": "# empty",
+        "src/streamlit/helpers/file2.py": "# empty",
+    }
+    with temp_local_dir(project_files) as project_root:
+        deploy_root = project_root / "output" / "deploy"
+        yield BundleMap(project_root=project_root, deploy_root=deploy_root)
+
+
+def ensure_path(path: Union[Path, str]) -> Path:
+    if isinstance(path, str):
+        return Path(path)
+    return path
+
+
+def verify_mappings(
+    bundle_map: BundleMap,
+    expected_mappings: Dict[Union[str, Path], Optional[Union[str, Path]]],
+    **kwargs,
+):
+    for src, dest in expected_mappings.items():
+        if dest is None:
+            assert bundle_map.to_deploy_path(ensure_path(src)) is None
+        else:
+            assert bundle_map.to_deploy_path(ensure_path(src)) == ensure_path(dest)
+
+    actual_path_mappings = dict(bundle_map.all_mappings(**kwargs))
+    expected_path_mappings = {
+        ensure_path(src): ensure_path(dest)
+        for (src, dest) in expected_mappings.items()
+        if dest is not None
+    }
+    assert actual_path_mappings == expected_path_mappings
+
+
+def path_mapping_factory(src: str, dest: Optional[str] = None) -> PathMapping:
+    return PathMapping(src=src, dest=dest)
+
+
+def test_empty_bundle_map(bundle_map):
+    mappings = list(bundle_map.all_mappings())
+    assert mappings == []
+
+    verify_mappings(
+        bundle_map,
+        {
+            "app/setup.sql": None,
+            ".": None,
+            "/not/in/project": None,
+        },
+    )
+
+
+@pytest.mark.parametrize("mapping_factory", [ArtifactMapping, path_mapping_factory])
```

Comment:
These parameterizations are just giving us a bit more test coverage because we're using `isinstance` instead of duck typing (are `src` and `dest` both strings?) in `BundleMap`. I'm not suggesting a change, just thinking aloud

Reply:
There are two entry points for add, at least conceptually. One uses a PathMapping as input, the other uses and ArtifactMapping. They're really pretty much equivalent, except that PathMapping carries around the optional `dest` (and soon processors). TBH I want to remove ArtifactMapping now that we moved over to pydantic. That class made sense when we were getting the original contents of the yaml file as a dictionary, but it's very redundant now. I wrote tests this way because removing ArtifactMapping would then be trivial :)

Reply:
That makes a ton of sense. Optimize for the clarity of the refactor you know is coming.